i18n-helpers/src/bin/markdown-gettext.rs

```diff
@@ -0,0 +1,250 @@
+            .with_context(|| format!("Failed to write to file: {}", output_file_path.display()))?;
+    }
+    Ok(())
+}
+
+fn validate_file(path: &Path) -> bool {
+    path.is_file() && path.extension().map_or(false, |ext| ext == "md")
+}
+
+fn allocate_files(
+    dir_option: Option<&str>,
+    file_option: Option<&str>,
+) -> anyhow::Result<Vec<PathBuf>> {
+    let mut valid_files = Vec::new();
+    match (dir_option, file_option) {
+        (Some(dir_path), None) => {
+            let full_dir = PathBuf::from(dir_path);
+            fs::read_dir(full_dir)?
+                .filter_map(Result::ok)
+                .filter(|entry| validate_file(&entry.path()))
+                .for_each(|entry| valid_files.push(entry.path()));
+        }
+        (None, Some(file_path)) => {
+            let full_file_path = PathBuf::from(file_path);
+            if !validate_file(&full_file_path) {
+                bail!("Markdown file does not exist: {}", full_file_path.display())
+            }
+            valid_files.push(full_file_path);
+        }
+        _ => unreachable!(),
+    }
+    Ok(valid_files)
+}
+
+fn build_output_path(output_path_option: Option<&str>) -> anyhow::Result<PathBuf> {
+    let output_path = output_path_option.unwrap_or("translated_md_files");
+    let path = PathBuf::from(output_path);
+    fs::create_dir_all(&path)
+        .with_context(|| format!("Failed to create directory: {}", path.display()))?;
+    Ok(path)
+}
+
+fn main() -> anyhow::Result<()> {
+    let matches = Command::new("markdown-gettext")
+        .about("markdown translator binary for mdbook-i18n-helpers")
+        .arg(Arg::new("input_dir").short('d').long("dir"))
+        .arg(Arg::new("file").short('f').long("file"))
+        .arg(Arg::new("po").short('p').long("po").required(true))
+        .arg(Arg::new("output_dir").short('o').long("out"))
+        .group(
+            ArgGroup::new("input_source")
+                .args(["input_dir", "file"])
+                .required(true)
+                .multiple(false),
+        )
+        .get_matches();
+
+    let lang = matches.get_one::<String>("po").unwrap().as_str();
+    let dir_option = matches.get_one::<String>("input_dir").map(|d| d.as_str());
+    let file_option = matches.get_one::<String>("file").map(|f| f.as_str());
+    let output_path_option = matches.get_one::<String>("output_dir").map(|d| d.as_str());
+
+    let files = allocate_files(dir_option, file_option)?;
+    let catalog = build_catalog(lang)?;
+    let output_path = build_output_path(output_path_option)?;
+
+    translate_files(catalog, files, output_path)?;
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use tempfile::TempDir;
+
+    static MARKDOWN_CONTENT: &str = "# How to Foo from a specified file path\n\
+        \n\
+        The first paragraph about Foo.\n\
+        Still the first paragraph.*baz*\n";
+
+    static INVALID_CONTENT: &str = "<p> This is not markdown content </p>";
```

Comment:
Actually... Markdown can contain HTML tags, so this string is also valid Markdown :slightly_smiling_face: